OpenDDS core doesn't compile with clang20 and C++23 enabled

See https://github.com/OpenDDS/OpenDDS/pull/5088, looks the core doesn't compile yet with clang20 and C++23

```
clang++-20 -std=c++23 -g   -O0 -DACE_HAS_CUSTOM_EXPORT_MACROS=0  -I/home/runner/work/OpenDDS/OpenDDS/ACE_TAO/ACE -DACE_NO_INLINE -I/home/runner/work/OpenDDS/OpenDDS/ACE_TAO/ACE -I../common  -c -o .obj/main.o /home/runner/work/OpenDDS/OpenDDS/OpenDDS/performance-tests/DCPS/TCPProfilingTest/raw_tcp/subscriber/main.cpp
In file included from /home/runner/work/OpenDDS/OpenDDS/OpenDDS/dds/DCPS/XTypes/TypeObject.cpp:7:
In file included from /home/runner/work/OpenDDS/OpenDDS/OpenDDS/dds/DCPS/XTypes/TypeObject.h:9:
In file included from /home/runner/work/OpenDDS/OpenDDS/OpenDDS/dds/DCPS/XTypes/External.h:9:
In file included from /home/runner/work/OpenDDS/OpenDDS/OpenDDS/dds/DCPS/unique_ptr.h:4:
In file included from /home/runner/work/OpenDDS/OpenDDS/OpenDDS/dds/DCPS/Definitions.h:15:
In file included from /home/runner/work/OpenDDS/OpenDDS/ACE_TAO/ACE/ace/Message_Block.h:25:
In file included from /home/runner/work/OpenDDS/OpenDDS/ACE_TAO/ACE/ace/Time_Value.h:25:
In file included from /usr/lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/chrono:49:
In file included from /usr/lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/shared_ptr.h:53:
In file included from /usr/lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/shared_ptr_base.h:59:
Error: /usr/lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/unique_ptr.h:91:16: error: invalid application of 'sizeof' to an incomplete type 'OpenDDS::XTypes::TypeIdentifier'
   91 |         static_assert(sizeof(_Tp)>0,
      |                       ^~~~~~~~~~~
/usr/lib/gcc/x86_64-linux-gnu/14/../../../../include/c++/14/bits/unique_ptr.h:398:4: note: in instantiation of member function 'std::default_delete<OpenDDS::XTypes::TypeIdentifier>::operator()' requested here
  398 |           get_deleter()(std::move(__ptr));
      |           ^
/home/runner/work/OpenDDS/OpenDDS/OpenDDS/dds/DCPS/XTypes/External.h:26:7: note: in instantiation of member function 'std::unique_ptr<OpenDDS::XTypes::TypeIdentifier>::~unique_ptr' requested here
   26 | class External {
      |       ^
/home/runner/work/OpenDDS/OpenDDS/OpenDDS/dds/DCPS/XTypes/TypeObject.h:409:5: note: in implicit destructor for 'OpenDDS::XTypes::External<OpenDDS::XTypes::TypeIdentifier>' first required here
  409 |     PlainSequenceSElemDefn()
      |     ^
/home/runner/work/OpenDDS/OpenDDS/OpenDDS/dds/DCPS/XTypes/TypeObject.h:341:9: note: forward declaration of 'OpenDDS::XTypes::TypeIdentifier'
  341 |   class TypeIdentifier;
      |         ^
make[1]: Entering directory '/home/runner/work/OpenDDS/OpenDDS/OpenDDS/tests/DCPS/TcpReconnect'
touch .depend.stub
1 error generated.
```

--- a/.github/workflows/build_and_test.yml
+++ b/.github/workflows/build_and_test.yml
@@ -1764,7 +1764,7 @@ jobs:
         cat "$GITHUB_WORKSPACE/OpenDDS/${{ github.job }}_autobuild_workspace/logs/latest.txt"
         grep -q 'Failures: 0' "$GITHUB_WORKSPACE/OpenDDS/${{ github.job }}_autobuild_workspace/logs/latest.txt"
 
-  ACE_TAO_u24_clang17_sec_js0:
+  ACE_TAO_u24_clang20_sec_js0:
 
     runs-on: ubuntu-24.04
 
@@ -1786,10 +1786,16 @@ jobs:
         cd OpenDDS/ACE_TAO
         export ACE_COMMIT=$(git rev-parse HEAD)
         echo "ACE_COMMIT=$ACE_COMMIT" >> $GITHUB_ENV
+    - name: add LLVM apt repo llvm-toolchain-noble-20
+      run: |
+        wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
+        sudo apt-add-repository "deb http://apt.llvm.org/noble/ llvm-toolchain-noble-20 main"
+    - name: install clang-20
+      run: sudo apt-get -y install clang-20
     - name: get compiler version
       shell: bash
       run: |
-        export COMPILER_VERSION=$(clang++-17 --version 2>&1 | head -n 1)
+        export COMPILER_VERSION=$(clang++-20 --version 2>&1 | head -n 1)
         echo "COMPILER_VERSION=$COMPILER_VERSION" >> $GITHUB_ENV
     - name: Cache Artifact
       id: cache-artifact
@@ -1813,7 +1819,7 @@ jobs:
         ./configure \
           --ace=$GITHUB_WORKSPACE/OpenDDS/ACE_TAO/ACE \
           --mpc=$GITHUB_WORKSPACE/MPC \
-          --compiler=clang++-17 \
+          --compiler=clang++-20 \
           --std=c++11 \
           --macros=SOFLAGS+=-Wl,-h,\$\(SONAME\) \
           --tests \
@@ -1845,11 +1851,11 @@ jobs:
         name: ${{ github.job }}_artifact
         path: ${{ github.job }}.tar.xz
 
-  build_u24_clang17_sec_js0:
+  build_u24_clang20_sec_js0:
 
     runs-on: ubuntu-24.04
 
-    needs: ACE_TAO_u24_clang17_sec_js0
+    needs: ACE_TAO_u24_clang20_sec_js0
 
     steps:
     - name: Update apt indexes
@@ -1870,26 +1876,32 @@ jobs:
     - name: download ACE_TAO artifact
       uses: actions/download-artifact@v5
       with:
-        name: ACE_TAO_u24_clang17_sec_js0_artifact
+        name: ACE_TAO_u24_clang20_sec_js0_artifact
         path: ACE_TAO
     - name: extract ACE_TAO artifact
       shell: bash
       run: |
         cd ACE_TAO
-        tar xvfJ ACE_TAO_u24_clang17_sec_js0.tar.xz
+        tar xvfJ ACE_TAO_u24_clang20_sec_js0.tar.xz
     - name: checkout OpenDDS
       uses: actions/checkout@v5
       with:
         path: OpenDDS
         submodules: true
+    - name: add LLVM apt repo llvm-toolchain-noble-20
+      run: |
+        wget -O - https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
+        sudo apt-add-repository "deb http://apt.llvm.org/noble/ llvm-toolchain-noble-20 main"
+    - name: install clang-20
+      run: sudo apt-get -y install clang-20
     - name: configure OpenDDS
       run: |
         cd OpenDDS
         ./configure \
           --ace=$GITHUB_WORKSPACE/ACE_TAO/ACE \
           --mpc=$GITHUB_WORKSPACE/MPC \
-          --compiler=clang++-17 \
-          --std=c++11 \
+          --compiler=clang++-20 \
+          --std=c++23 \
           --macros=SOFLAGS+=-Wl,-h,\$\(SONAME\) \
           --tests \
           --security \
@@ -1914,8 +1926,8 @@ jobs:
         source setenv.sh
         cmake -S . -B build \
           -DOPENDDS_JUST_BUILD_TESTS=TRUE \
-          -DCMAKE_CXX_STANDARD=11 \
-          -DCMAKE_CXX_COMPILER=clang++-17
+          -DCMAKE_CXX_STANDARD=23 \
+          -DCMAKE_CXX_COMPILER=clang++-20
         cmake --build build -- -j4
     - name: create OpenDDS tar.xz artifact
       shell: bash
@@ -1932,11 +1944,11 @@ jobs:
         name: ${{ github.job }}_artifact
         path: ${{ github.job }}.tar.xz
 
-  cmake_u24_clang17_sec_js0:
+  cmake_u24_clang20_sec_js0:
 
     runs-on: ubuntu-24.04
 
-    needs: build_u24_clang17_sec_js0
+    needs: build_u24_clang20_sec_js0
 
     permissions:
       contents: read
@@ -1966,13 +1978,13 @@ jobs:
     - name: download ACE_TAO artifact
       uses: actions/download-artifact@v5
       with:
-        name: ACE_TAO_u24_clang17_sec_js0_artifact
+        name: ACE_TAO_u24_clang20_sec_js0_artifact
         path: ACE_TAO
     - name: extract ACE_TAO artifact
       shell: bash
       run: |
         cd ACE_TAO
-        tar xvfJ ACE_TAO_u24_clang17_sec_js0.tar.xz
+        tar xvfJ ACE_TAO_u24_clang20_sec_js0.tar.xz
     - name: checkout OpenDDS
       uses: actions/checkout@v5
       with:
@@ -1981,13 +1993,13 @@ jobs:
     - name: download OpenDDS artifact
       uses: actions/download-artifact@v5
       with:
-        name: build_u24_clang17_sec_js0_artifact
+        name: build_u24_clang20_sec_js0_artifact
         path: OpenDDS
     - name: extract OpenDDS artifact
       shell: bash
       run: |
         cd OpenDDS
-        tar xvfJ build_u24_clang17_sec_js0.tar.xz
+        tar xvfJ build_u24_clang20_sec_js0.tar.xz
     - name: check build configuration
       shell: bash
       run: |

--- a/.github/workflows/build_and_test.yml
+++ b/.github/workflows/build_and_test.yml
@@ -1802,7 +1802,7 @@ jobs:
       uses: actions/cache@v4
       with:
         path: ${{ github.job }}.tar.xz
-        key: c02_${{ github.job }}_ace6tao2_${{ env.ACE_COMMIT }}_${{ env.COMPILER_VERSION }}
+        key: c03_${{ github.job }}_ace6tao2_${{ env.ACE_COMMIT }}_${{ env.COMPILER_VERSION }}
     - name: install xerces
       if: steps.cache-artifact.outputs.cache-hit != 'true'
       run: sudo apt-get -y install libxerces-c-dev
@@ -1820,7 +1820,7 @@ jobs:
           --ace=$GITHUB_WORKSPACE/OpenDDS/ACE_TAO/ACE \
           --mpc=$GITHUB_WORKSPACE/MPC \
           --compiler=clang++-20 \
-          --std=c++11 \
+          --std=c++23 \
           --macros=SOFLAGS+=-Wl,-h,\$\(SONAME\) \
           --tests \
           --security \

--- a/configure
+++ b/configure
@@ -744,6 +744,16 @@ sub ndk_clang {
 
 ## compiler
 
+sub get_compiler_type {
+  my $comp = shift;
+  return 'MSVC' if $comp =~ /cl(\.exe)?$/i;
+  return 'Clang' if $comp =~ /clang/;
+  return 'GNU' if $comp =~ /g\+\+/;
+  return undef;
+}
+
+my $compiler_type = undef;
+
 if ($has_host_compiler) {
   if ($opts{'compiler'}) {
     my $standard = 0;
@@ -769,9 +779,11 @@ if ($has_host_compiler) {
                                   "Command Prompt.\n" : '') . "Stopped";
     }
   }
-  print "compiler is: $opts{'compiler'}\n" if $opts{'verbose'};
 
-  if ($opts{'compiler'} =~ /cl(\.exe)?$/i) {
+  $compiler_type = get_compiler_type($opts{'compiler'});
+  print "compiler is: $opts{'compiler'}; type $compiler_type\n" if $opts{'verbose'};
+
+  if ($compiler_type eq 'MSVC') {
     my $savepath = $ENV{'PATH'};
     my $clpath = which($opts{'compiler'});
     if ($clpath) {
@@ -899,7 +911,7 @@ EOF
       }
     }
   }
-  elsif ($opts{'compiler'} =~ /g\+\+|clang/) {
+  elsif ($compiler_type eq 'Clang' || $compiler_type eq 'GNU') {
     my $version_string = `$opts{'compiler'} --version`;
     print "Compiler version: $version_string\n" if $opts{'verbose'};
     $opts{'is apple clang'} = $version_string =~ /^Apple/;
@@ -918,7 +930,7 @@ EOF
         print "Detected GCC >= 6, default -std=gnu++14\n" if $opts{'verbose'};
       }
     }
-    elsif ($opts{'compiler'} =~ /clang/ && $version_string !~ /^Apple / &&
+    elsif ($compiler_type eq 'Clang' && $version_string !~ /^Apple / &&
            $version_string =~ /(\d+)\.(\d+)\.(\d+)/) {
       # Apple's version of Clang doesn't default to c++11 or higher, users
       # can pass --std= to change the C++ standard version used.
@@ -940,10 +952,9 @@ EOF
 my $compile_warnings = $opts{'compile-warnings'} // '';
 if ($compile_warnings eq 'WARNING' or $compile_warnings eq 'ERROR') {
   my ($section_names, $sections) = read_ini_file("$FindBin::RealBin/build.ini");
-  my %compiler_map = ( 'g++' => 'GNU', 'clang' => 'Clang', 'cl' => 'MSVC', 'cl.exe' => 'MSVC');
-  my $section = $compiler_map{$opts{'compiler'}};
-  if ($opts{'is apple clang'}) {
-    $section = 'AppleClang';
+  my $section = $opts{'is apple clang'} ? 'AppleClang' : $compiler_type;
+  unless (exists $sections->{$section}) {
+    die "Can't determine compiler type to enable warnings.\n Stopped";
   }
   my $warning_flags = $sections->{$section}{'warning'};
   my $error_flags = $sections->{$section}{'error'};

--- a/dds/DCPS/XTypes/TypeObject.cpp
+++ b/dds/DCPS/XTypes/TypeObject.cpp
@@ -35,6 +35,86 @@ DCPS::String equivalence_hash_to_string(const EquivalenceHash& equivalence_hash)
   return DCPS::to_dds_string(&equivalence_hash[0], sizeof equivalence_hash);
 }
 
+PlainSequenceSElemDefn::PlainSequenceSElemDefn()
+  : bound(INVALID_SBOUND)
+{}
+
+PlainSequenceSElemDefn::PlainSequenceSElemDefn(const PlainCollectionHeader& a_header,
+                                               const SBound& a_bound,
+                                               const TypeIdentifier& a_element_identifier)
+  : header(a_header)
+  , bound(a_bound)
+  , element_identifier(a_element_identifier)
+{}
+
+PlainSequenceLElemDefn::PlainSequenceLElemDefn()
+  : bound(INVALID_LBOUND)
+{}
+
+PlainSequenceLElemDefn::PlainSequenceLElemDefn(const PlainCollectionHeader& a_header,
+                                               const LBound& a_bound,
+                                               const TypeIdentifier& a_element_identifier)
+  : header(a_header)
+  , bound(a_bound)
+  , element_identifier(a_element_identifier)
+{}
+
+PlainArraySElemDefn::PlainArraySElemDefn()
+{}
+
+PlainArraySElemDefn::PlainArraySElemDefn(const PlainCollectionHeader& a_header,
+                                         const SBoundSeq& a_array_bound_seq,
+                                         const TypeIdentifier& a_element_identifier)
+  : header(a_header)
+  , array_bound_seq(a_array_bound_seq)
+  , element_identifier(a_element_identifier)
+{}
+
+PlainArrayLElemDefn::PlainArrayLElemDefn()
+{}
+
+PlainArrayLElemDefn::PlainArrayLElemDefn(const PlainCollectionHeader& a_header,
+                                         const LBoundSeq& a_array_bound_seq,
+                                         const TypeIdentifier& a_element_identifier)
+  : header(a_header)
+  , array_bound_seq(a_array_bound_seq)
+  , element_identifier(a_element_identifier)
+{}
+
+PlainMapSTypeDefn::PlainMapSTypeDefn()
+  : bound(INVALID_SBOUND)
+  , key_flags(0)
+{}
+
+PlainMapSTypeDefn::PlainMapSTypeDefn(const PlainCollectionHeader& a_header,
+                                     const SBound& a_bound,
+                                     const TypeIdentifier& a_element_identifier,
+                                     const CollectionElementFlag& a_key_flags,
+                                     const TypeIdentifier& a_key_identifier)
+  : header(a_header)
+  , bound(a_bound)
+  , element_identifier(a_element_identifier)
+  , key_flags(a_key_flags)
+  , key_identifier(a_key_identifier)
+{}
+
+PlainMapLTypeDefn::PlainMapLTypeDefn()
+  : bound(INVALID_LBOUND)
+  , key_flags(0)
+{}
+
+PlainMapLTypeDefn::PlainMapLTypeDefn(const PlainCollectionHeader& a_header,
+                                     const LBound& a_bound,
+                                     const TypeIdentifier& a_element_identifier,
+                                     const CollectionElementFlag& a_key_flags,
+                                     const TypeIdentifier& a_key_identifier)
+  : header(a_header)
+  , bound(a_bound)
+  , element_identifier(a_element_identifier)
+  , key_flags(a_key_flags)
+  , key_identifier(a_key_identifier)
+{}
+
 MinimalMemberDetail::MinimalMemberDetail(const OPENDDS_STRING& name)
 {
   unsigned char result[16];

--- a/dds/DCPS/XTypes/TypeObject.h
+++ b/dds/DCPS/XTypes/TypeObject.h
@@ -401,7 +401,7 @@ namespace XTypes {
     }
   };
 
-  struct PlainSequenceSElemDefn {
+  struct OpenDDS_Dcps_Export PlainSequenceSElemDefn {
     PlainCollectionHeader header;
     SBound bound;
     External<TypeIdentifier> element_identifier;
@@ -415,7 +415,7 @@ namespace XTypes {
     bool operator<(const PlainSequenceSElemDefn& other) const;
   };
 
-  struct PlainSequenceLElemDefn {
+  struct OpenDDS_Dcps_Export PlainSequenceLElemDefn {
     PlainCollectionHeader header;
     LBound bound;
     External<TypeIdentifier> element_identifier;
@@ -429,7 +429,7 @@ namespace XTypes {
     bool operator<(const PlainSequenceLElemDefn& other) const;
   };
 
-  struct PlainArraySElemDefn {
+  struct OpenDDS_Dcps_Export PlainArraySElemDefn {
     PlainCollectionHeader header;
     SBoundSeq array_bound_seq;
     External<TypeIdentifier> element_identifier;
@@ -443,7 +443,7 @@ namespace XTypes {
     bool operator<(const PlainArraySElemDefn& other) const;
   };
 
-  struct PlainArrayLElemDefn {
+  struct OpenDDS_Dcps_Export PlainArrayLElemDefn {
     PlainCollectionHeader header;
     LBoundSeq array_bound_seq;
     External<TypeIdentifier> element_identifier;
@@ -457,7 +457,7 @@ namespace XTypes {
     bool operator<(const PlainArrayLElemDefn& other) const;
   };
 
-  struct PlainMapSTypeDefn {
+  struct OpenDDS_Dcps_Export PlainMapSTypeDefn {
     PlainCollectionHeader header;
     SBound bound;
     External<TypeIdentifier> element_identifier;
@@ -475,7 +475,7 @@ namespace XTypes {
     bool operator<(const PlainMapSTypeDefn& other) const;
   };
 
-  struct PlainMapLTypeDefn {
+  struct OpenDDS_Dcps_Export PlainMapLTypeDefn {
     PlainCollectionHeader header;
     LBound bound;
     External<TypeIdentifier> element_identifier;

--- a/dds/DCPS/XTypes/TypeObject.h
+++ b/dds/DCPS/XTypes/TypeObject.h
@@ -406,17 +406,11 @@ namespace XTypes {
     SBound bound;
     External<TypeIdentifier> element_identifier;
 
-    PlainSequenceSElemDefn()
-      : bound(INVALID_SBOUND)
-    {}
+    PlainSequenceSElemDefn();
 
     PlainSequenceSElemDefn(const PlainCollectionHeader& a_header,
                            const SBound& a_bound,
-                           const TypeIdentifier& a_element_identifier)
-      : header(a_header)
-      , bound(a_bound)
-      , element_identifier(a_element_identifier)
-    {}
+                           const TypeIdentifier& a_element_identifier);
 
     bool operator<(const PlainSequenceSElemDefn& other) const;
   };
@@ -426,17 +420,11 @@ namespace XTypes {
     LBound bound;
     External<TypeIdentifier> element_identifier;
 
-    PlainSequenceLElemDefn()
-      : bound(INVALID_LBOUND)
-    {}
+    PlainSequenceLElemDefn();
 
     PlainSequenceLElemDefn(const PlainCollectionHeader& a_header,
                            const LBound& a_bound,
-                           const TypeIdentifier& a_element_identifier)
-      : header(a_header)
-      , bound(a_bound)
-      , element_identifier(a_element_identifier)
-    {}
+                           const TypeIdentifier& a_element_identifier);
 
     bool operator<(const PlainSequenceLElemDefn& other) const;
   };
@@ -446,15 +434,11 @@ namespace XTypes {
     SBoundSeq array_bound_seq;
     External<TypeIdentifier> element_identifier;
 
-    PlainArraySElemDefn() {}
+    PlainArraySElemDefn();
 
     PlainArraySElemDefn(const PlainCollectionHeader& a_header,
                         const SBoundSeq& a_array_bound_seq,
-                        const TypeIdentifier& a_element_identifier)
-      : header(a_header)
-      , array_bound_seq(a_array_bound_seq)
-      , element_identifier(a_element_identifier)
-    {}
+                        const TypeIdentifier& a_element_identifier);
 
     bool operator<(const PlainArraySElemDefn& other) const;
   };
@@ -464,15 +448,11 @@ namespace XTypes {
     LBoundSeq array_bound_seq;
     External<TypeIdentifier> element_identifier;
 
-    PlainArrayLElemDefn() {}
+    PlainArrayLElemDefn();
 
     PlainArrayLElemDefn(const PlainCollectionHeader& a_header,
                         const LBoundSeq& a_array_bound_seq,
-                        const TypeIdentifier& a_element_identifier)
-      : header(a_header)
-      , array_bound_seq(a_array_bound_seq)
-      , element_identifier(a_element_identifier)
-    {}
+                        const TypeIdentifier& a_element_identifier);
 
     bool operator<(const PlainArrayLElemDefn& other) const;
   };
@@ -484,22 +464,13 @@ namespace XTypes {
     CollectionElementFlag key_flags;
     External<TypeIdentifier> key_identifier;
 
-    PlainMapSTypeDefn()
-      : bound(INVALID_SBOUND)
-      , key_flags(0)
-    {}
+    PlainMapSTypeDefn();
 
     PlainMapSTypeDefn(const PlainCollectionHeader& a_header,
                       const SBound& a_bound,
                       const TypeIdentifier& a_element_identifier,
                       const CollectionElementFlag& a_key_flags,
-                      const TypeIdentifier& a_key_identifier)
-      : header(a_header)
-      , bound(a_bound)
-      , element_identifier(a_element_identifier)
-      , key_flags(a_key_flags)
-      , key_identifier(a_key_identifier)
-    {}
+                      const TypeIdentifier& a_key_identifier);
 
     bool operator<(const PlainMapSTypeDefn& other) const;
   };
@@ -511,22 +482,13 @@ namespace XTypes {
     CollectionElementFlag key_flags;
     External<TypeIdentifier> key_identifier;
 
-    PlainMapLTypeDefn()
-      : bound(INVALID_LBOUND)
-      , key_flags(0)
-    {}
+    PlainMapLTypeDefn();
 
     PlainMapLTypeDefn(const PlainCollectionHeader& a_header,
                       const LBound& a_bound,
                       const TypeIdentifier& a_element_identifier,
                       const CollectionElementFlag& a_key_flags,
-                      const TypeIdentifier& a_key_identifier)
-      : header(a_header)
-      , bound(a_bound)
-      , element_identifier(a_element_identifier)
-      , key_flags(a_key_flags)
-      , key_identifier(a_key_identifier)
-    {}
+                      const TypeIdentifier& a_key_identifier);
 
     bool operator<(const PlainMapLTypeDefn& other) const;
   };

--- a/dds/idl/langmap_generator.cpp
+++ b/dds/idl/langmap_generator.cpp
@@ -1524,12 +1524,19 @@ struct Cxx11Generator : GeneratorBase {
       if (af.cls_ & CL_ARRAY) {
         initializer = "{}";
       }
-      be_global->add_include("<utility>", BE_GlobalData::STREAM_LANG_H);
+      be_global->add_include("<utility>", BE_GlobalData::STREAM_CPP);
       be_global->lang_header_ <<
-        "  void " << af.name_ << "(const " << lang_field_type << "& val) " << assign <<
-        "  void " << af.name_ << '(' << lang_field_type << "&& val) " << move <<
+        "  void " << af.name_ << "(const " << lang_field_type << "& val);\n"
+        "  void " << af.name_ << '(' << lang_field_type << "&& val);\n"
         "  const " << lang_field_type << "& " << af.name_ << "() const " << ret <<
         "  " << lang_field_type << "& " << af.name_ << "() " << ret;
+      AST_Decl* const scope = dynamic_cast<AST_Decl*>(field->defined_in());
+      const char* const clazz = scope ? scope->local_name()->get_string() : "";
+      be_global->impl_ <<
+        "void " << clazz << "::" << af.name_ << "(const " << lang_field_type << "& val)\n"
+        << assign << "\n"
+        "void " << clazz << "::" << af.name_ << '(' << lang_field_type << "&& val)\n"
+        << move << "\n";
     }
     be_global->lang_header_ <<
       "  " << lang_field_type << " _" << af.name_ << initializer << ";\n\n";
@@ -1658,15 +1665,15 @@ struct Cxx11Generator : GeneratorBase {
       dval = strm.str();
     }
 
-    std::string disc_param, disc_name = dval;
+    std::string disc_param, disc_param_no_default, disc_name = dval;
     if (label->label_kind() == AST_UnionLabel::UL_default ||
         branch->label_list_length() > 1) {
       disc_name = "disc";
-      disc_param = ", " + disc_type + " disc = " + dval;
+      disc_param_no_default = ", " + disc_type + " disc";
+      disc_param = disc_param_no_default + " = " + dval;
     }
 
-    const std::string assign_pre = "{ _activate(" + disc_name + "); _"
-      + std::string(nm) + " = ",
+    const std::string assign_pre = "{ _activate(" + disc_name + "); _" + std::string(nm) + " = ",
       assign = assign_pre + "val; }\n",
       move = assign_pre + "std::move(val); }\n",
       ret = "{ return _" + std::string(nm) + "; }\n";
@@ -1677,14 +1684,19 @@ struct Cxx11Generator : GeneratorBase {
         "  " << lang_field_type << ' ' << nm << "() const " << ret <<
         "  " << lang_field_type << "& " << nm << "() " << ret << "\n";
     } else {
-      be_global->add_include("<utility>", BE_GlobalData::STREAM_LANG_H);
+      be_global->add_include("<utility>", BE_GlobalData::STREAM_CPP);
       be_global->lang_header_ <<
-        "  void " << nm << "(const " << lang_field_type << "& val" << disc_param
-        << ") " << assign <<
-        "  void " << nm << '(' << lang_field_type << "&& val" << disc_param
-        << ") " << move <<
+        "  void " << nm << "(const " << lang_field_type << "& val" << disc_param << ");\n"
+        "  void " << nm << '(' << lang_field_type << "&& val" << disc_param << ");\n"
         "  const " << lang_field_type << "& " << nm << "() const " << ret <<
         "  " << lang_field_type << "& " << nm << "() " << ret << "\n";
+      AST_Decl* const scope = dynamic_cast<AST_Decl*>(branch->defined_in());
+      const char* const clazz = scope ? scope->local_name()->get_string() : "";
+      be_global->impl_ <<
+        "void " << clazz << "::" << nm << "(const " << lang_field_type << "& val" << disc_param_no_default << ")\n"
+        << assign << "\n"
+        "void " << clazz << "::" << nm << '(' << lang_field_type << "&& val" << disc_param_no_default << ")\n"
+        << move << "\n";
     }
   }
 

